documentation/wiki/Binary-Log.md

```diff
@@ -94,4 +94,96 @@ When incrementing the file format, keep this in mind:
+
+We recommend usage of `BinaryLogReplayEventSource`. It provides simplified helpers for creating and configuring `BuildEventArgsReader` and subscribing to the events.
+
+```csharp
+var logReader = new BinaryLogReplayEventSource()
+{
+    AllowForwardCompatibility = true
+};
+
+// Handling of the structured events contained within the log
+logReader.AnyEventRaised += (_, e) =>
+{
+    if (e is BuildErrorEventArgs error)
+    {
+        //...
+    }
+
+    // ...
+};
+
+// Starts the synchronous log reading loop.
+logReader.Replay(path_to_binlog_file);
+
+```
+
+### Handling the recoverable reading errors
+
+In compatibility mode (default for `BinaryLogReplayEventSource`. Only supported for binlogs of version 18 and higher) reader is capable of skipping unknown event types and unknown parts of known events (`BuildEventArgsReader` can configure the behavior via 2 separate properties - `SkipUnknownEvents` and `SkipUnknownEventParts`).
+
+The unknown events and event parts are regarded as recoverable errors, since the reader is able to continue reading subsequent records in the binlog. However the specific user logic should have the last call in deciding whether errors are really recoverable (e.g. is presence of unrecognized or unparseable event ok? It might be fine when searching only for specific events - e.g. errors but not acceptable when trying to provide definitive overview of the built).
+
+To allow the calling code to decide - based on the type of error, type of events getting the error, or the number of errors - the `OnRecoverableReadError` event is exposed (from both `BinaryLogReplayEventSource` and `BuildEventArgsReader`).
+
+```csharp
+/// <summary>
+/// An event args for <see cref="IBinaryLogReaderErrors.OnRecoverableReadError"/> event.
+/// </summary>
+public sealed class BinaryLogReaderErrorEventArgs : EventArgs
+{
+    /// <summary>
+    /// Type of the error that occurred during reading.
+    /// </summary>
+    public ReaderErrorType ErrorType { get; }
+
+    /// <summary>
+    /// Kind of the record that encountered the error.
+    /// </summary>
+    public BinaryLogRecordKind RecordKind { get; }
+
+    /// <summary>
+    /// Materializes the error message.
+    /// Until it's called the error message is not materialized and no string allocations are made.
+    /// </summary>
+    /// <returns>The error message.</returns>
+    public string GetFormattedMessage() => _formatErrorMessage();
+}
+
+/// <summary>
+/// Receives recoverable errors during reading.
+/// Communicates type of the error, kind of the record that encountered the error and the message detailing the error.
+/// </summary>
+event Action<BinaryLogReaderErrorEventArgs>? OnRecoverableReadError;
+```
+
+Our sample usage of the [Reading API](#reading-api) can be enhanced with recoverable errors handling e.g. as such:
+
+```csharp
+
+// Those can be raised only during forward compatibility reading mode.
+logReader.OnRecoverableReadError += errorEventArgs =>
```

Comment:
Is it legal to fail to handle this? Or should "you didn't say you are willing to deal with a new binlog format by skipping data" be a fatal error?

Reply:
I'd say it's OK not to handle this.

In many scenarios the errors *should* be handled when fwd compatibility is opted in ... but possibly not in all scenarios.

I can imagine a case where user opts into forward compatibility (either via `BuildEventArgsReader.SkipUnknownEvents`/`BuildEventArgsReader.SkipUnknownEventParts` or `BinaryLogReplayEventSource.AllowForwardCompatibility`) in order to be able to ignore errors that they *are not interrested in* - they e.g. might just want to convert binlog to other format and feed it into some automated analysis.

Thoughts?

Reply:
I think my concern is: suppose we decide in some future version that we need a new event for something fundamental, in order to better represent its data. So now we never emit (for example) `TaskStartedEventsArgs` and only emit `BetterTaskStartedEventsArgs`. That's not a _breaking_ change, and some binlog parsers may never care. But if you're analyzing task invocations you certainly would!

Feels like a case where we should make it a fatal error now because it's easy to reverse that, and hard to go the other way.

Reply:
That makes sense - I'm adding the check.
I wanted to add the check to the fwd compatibility reading opt-in properties, which would prevent repetitive checking in `Read()` - but that would mean that those property initializers would not be possible to use (as event initialization during object initialization is not yet possible: https://github.com/dotnet/csharplang/issues/5176)
For this reason it's part of `Read()`, but it should short circuit quickly and not affect perf in a measurable way.